libdrgn/python/program.c

```diff
@@ -481,6 +481,89 @@ static struct drgn_error *py_object_find_fn(const char *name, size_t name_len,
 	return drgn_object_copy(ret, &((DrgnObject *)obj)->obj);
 }
 
+static struct drgn_error *py_symbol_find_fn(const char *name, uint64_t addr,
+					    struct drgn_module *module,
+					    enum drgn_find_symbol_flags flags,
+					    void *data, struct drgn_symbol_result_builder *builder)
+{
+	struct drgn_error *err = NULL;
+	PyGILState_STATE gstate;
+	_cleanup_pydecref_ PyObject *name_obj = NULL;
+	_cleanup_pydecref_ PyObject *address_obj = NULL;
+	_cleanup_pydecref_ PyObject *one_obj = NULL;
+	_cleanup_pydecref_ PyObject *obj = NULL;
+
+	gstate = PyGILState_Ensure();
+
+	if (flags & DRGN_FIND_SYM_NAME) {
+		name_obj = PyUnicode_FromString(name);
+		if (!name_obj) {
+			err = drgn_error_from_python();
+			goto out;
+		}
+	} else {
+		name_obj = Py_None;
+		Py_INCREF(name_obj);
+	}
+
+	if (flags & DRGN_FIND_SYM_ADDR) {
+		address_obj = PyLong_FromUnsignedLong(addr);
+		if (!address_obj) {
+			err = drgn_error_from_python();
+			goto out;
+		}
+	} else {
+		address_obj = Py_None;
+		Py_INCREF(address_obj);
+	}
+
+	one_obj = PyBool_FromLong(flags & DRGN_FIND_SYM_ONE);
+
+	obj = PyObject_CallFunction(data, "OOO", name_obj, address_obj, one_obj);
+	if (!obj) {
+		err = drgn_error_from_python();
+		goto out;
+	}
+	if (!PyList_Check(obj)) {
+		PyErr_SetString(PyExc_TypeError,
+				"symbol find callback must return list");
+		err = drgn_error_from_python();
+		goto out;
+	}
+
+	size_t len = PyList_GET_SIZE(obj);
+	if (len > 1 && (flags & DRGN_FIND_SYM_ONE))  {
+		PyErr_SetString(PyExc_ValueError,
+				"symbol find callback returned multiple elements, but one was requested");
+		err = drgn_error_from_python();
```

Comment:
I don't know what exactly I thought would happen as a result of this odd pattern, but what actually happens is:
```
Traceback (most recent call last):
  File "/home/stepbren/repos/drgn/tests/test_symbol.py", line 319, in test_args_single_with_many_results
    self.prog.symbol(0xFFFF2008)
Exception: ValueError: symbol find callback returned multiple elements, but one was requested
```
I'm going to directly return a drgn error here, and add a test covering this case.